### ISSUE TITLE
Job enqueued but never performing

### ISSUE THREAD

**Issue by LeoGardel** [collectiveidea/delayed_job#1041]
I'm using DJ in my Rails app. All my functions that are being enqueued are working fine, always using the method "handle_asynchronously", except one. This one works perfectly in development, but in my production environment (Heroku), it fails 50% of the times.

Debugging the problem, I found ou that the job is always being enqueued, it is being taken by a worker to be executed ('locked_at' and 'locked_by' are being filled), but the perform function only starts 50% of the times, randomly. Although, 100% of the times the job execution is considered to be successful and the line is deleted (i'm already keeping the failed jobs).

Any idea?? I'm running out of solutions and almost deciding to rewrite everything to ActiveJob patterns in order to get more flexibility to try different libraries, but I really want to understand what's going on and keep using DJ (mainly because I don't want to configure Redis now).

Using delayed_job 4.1.5
Using delayed_job_active_record 4.1.3
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
Rails 5.2.0

UPDATE -----

Below is my delayed method. I know it is not being called for two reasons: it prints a big text in the console and it creates a file on Amazon S3. I've already tried to add an "else" clause to the "unless" statement to check if it was failing the unless test, but it never got in there too.

```
  after_create :create_shareable_file

  def create_shareable_file
    unless Rails.env.test?
      puts "\n\n\n!!!!!\n\nStarting Amazon Upload\n\n!!!!!\n\n\n"
      uploaded = Aws::S3::Resource.new(region: ENV['S3_REGION'])
        .bucket(ENV['S3_BUCKET'])
        .object(final_path)
        .put(body: get_shareable_file.html_safe, cache_control: "max-age=0", content_type: "text/html")
      puts "\n\n\n!!!!!\n\nAmazon Upload Finished: #{uploaded}\n\n!!!!!\n\n\n"
    end
  end
  handle_asynchronously :create_shareable_file, queue: 'SHAREABLE.POSTS'
```

**Comment by albus522** [comment]
How are you verifying that perform is only called some of the time?

**Comment by LeoGardel** [comment]
I've just updated the post with your answer. :)

**Comment by LeoGardel** [comment]
David, I simply paused this task to accomplish another one: upgrade my Heroku DB to a paid one (it exceded 10k rows and they were warning me that I had to upgrade). I've made their super simple "pg:copy" method and did it. Now, all this async task is always working in production. I don't know if it was some kind of restart, but the problem seems to be misteriously solved, with the exact same code...

I'll post here if the problem starts happening again. Still interested on answering your questions about my case and on your thoughts about possible reasons for this to happen.

Just to recap:
- Always enqueues successfully
- Always fills "locked_at" and "locked_by" the job after a small period of time ("handle_async")
- Sometimes it executed the code, sometimes not
- Always destroys the job, even with `Delayed::Worker.destroy_failed_jobs = false`

Thanks
